instrumentation/bunny/lib/opentelemetry/instrumentation/bunny/patches/queue.rb

```diff
@@ -0,0 +1,36 @@
+# frozen_string_literal: true
+
+# Copyright The OpenTelemetry Authors
+#
+# SPDX-License-Identifier: Apache-2.0
+
+module OpenTelemetry
+  module Instrumentation
+    module Bunny
+      module Patches
+        # The Queue module contains the instrumentation patch the Queue#pop method.
+        module Queue
+          def pop(opts = { manual_ack: false }, &block)
+            super do |delivery_info, properties, payload|
+              if block
+                OpenTelemetry::Instrumentation::Bunny::PatchHelpers.with_process_span(tracer, delivery_info, properties) do
+                  yield delivery_info, properties, payload
+                end
+              else
+                OpenTelemetry::Instrumentation::Bunny::PatchHelpers.with_receive_span(tracer, delivery_info, properties) do
+                  [delivery_info, properties, payload]
+                end
+              end
+            end
+          end
```

Comment:
Looking at the code for `Queue#pop`, I don't think this is quite correct. I think we want to instrument `process` here only if a block is provided, and we should instrument `Channel#basic_get` to generate the `receive` span. This will result in `enqueue -> receive` in all cases, and `enqueue -> receive -> process` if a user calls `queue.pop { ... }`.

The challenge here is how to propagate the `span_id` from the `receive` span to use as the parent of the `process` span. I'd love to use `OpenTelemetry::Context` for this, but we should really only use this in a block-structured fashion. One option is to set a thread-local variable to record the last `receive` span ID. An alternative is to mutate the `properties[:header]` by injecting the `receive` span context.

Reply:
That makes sense, I used header propagation as with consumers the thread receiving the message is different from the thread processing the message.